### ISSUE TITLE
[ENHANCEMENT] correct error handling in initImapStream

### ISSUE THREAD

**Issue by WHMCSLawrence** [barbushin/php-imap#394]
**Environment (please complete the following information):**
 - PHP IMAP version: unsure
 - PHP Version: 7.2.23
 - Type of execution: CLI but doesn't matter

**Describe the bug**
In the Mailbox.php file, the initImapStream function fails to clear imap errors and doesn't properly handle connection errors (it assumes everything is an imap error). When used in a script that loops through reading multiple mailboxes inside a try/catch, an imap error to the first (for example using an invalid password, resulting in an authentication error) can be reported as the same error to the second if it has a non-imap error (for example: failure due to a firewall issue).

**To Reproduce**
Use the Mailbox class in a script that loops through multiple mailboxes and ensure that the above scenario occurs

The used code:
```Unable to provide the code unfortunately.
```

**Expected behavior**
Clear the imap error stack immediately before the return in the function and check that $lastError is set with an imap error. If it is, report the imap error. Otherwise we assume it is a connectivity issue, and report it as such.

Something like this should work (and works when I test it locally):

```
protected function initImapStream() {		
		$imapStream = @imap_open($this->imapPath, $this->imapLogin, $this->imapPassword, $this->imapOptions, $this->imapRetriesNum, $this->imapParams);
		if(!$imapStream) {
			$lastError = imap_last_error();
			
			if(!empty($lastError)){
				// imap error = report imap error
				throw new Exception("IMAP error: " . $lastError);
			}
			else {
				// no imap error = connectivity issue
				throw new Exception("Connection error: Unable to connect to " . $this->imapPath);
			}
		}
		
		// this function is called multiple times and imap keeps errors around. Let's clear them out to avoid it tripping up future calls.
		@imap_errors();
		
		return $imapStream;
	}
```

**Comment by Sebi94nbg** [comment]
@WHMCSLawrence your code to reproduce the issue is missing. Can you please provide it, that I can implement and test your suggested solution? Thanks in advance!

**Comment by WHMCSLawrence** [comment]
I am afraid that I cannot, as the code in question isn't permitted for redistribution. However the issue in your code is simple to reproduce: attempt to read 2 mailboxes in the same PHP script run inside a loop using try catch and have the first mailbox be empty. Because your code doesn't clear the "Mailbox is empty" message from IMAP, it can be mistakenly interpreted as a "Connection Error" in your current code. The new if statement is intended to help the user know whether it is an IMAP error or a general connection error (firewall for example).

**Comment by Sebi94nbg** [comment]
Alright, thanks for the feedback. I'll check this out.

**Comment by Sebi94nbg** [comment]
Committed and tested. Will be included in the next release. Thanks for your feedback and the suggestion!

**Comment by Sebi94nbg** [comment]
Released in version [3.0.31](https://github.com/barbushin/php-imap/releases/tag/3.0.31).
